src/wikigraph.cc

```diff
@@ -190,69 +194,93 @@ std::map<std::string, double> WikiGraph::getCentralityMap() const {
-    for (const auto& page : pages) predecessor[page] = {};
-    std::map<std::string, double> sigma;
-    for (const auto& page : pages) sigma[page] = 0.0;
-    sigma[start] = 1.0;
-    std::map<std::string, int> dist;
-    for (const auto& page : pages) dist[page] = -1;
-    dist[start] = 0;
-    std::queue<std::string> Q;
-    Q.push(start);
-    while(!Q.empty()) {
-      std::string v = Q.front(); Q.pop();
-      S.push(v);
-      for (const auto& w : article_map.at(v)) {
-        if (dist[w] < 0) {
-          Q.push(w);
-          dist[w] = dist[v] + 1;
-        }
-        if (dist[w] == dist[v] + 1) {
-          sigma[w] = sigma[w] + sigma[v];
-          predecessor[w].push_back(v);
-        }
-      }
-    }
+  return centralilty_map; // Default value for now
+}
+
+void WikiGraph::threadHelper(SafeQueue& q, std::map<std::string, double>& centrality_map, const std::vector<std::string>& pages) const {
+  std::optional<std::string> curr_page = q.pop();
+  while (curr_page.has_value()) {
+    brandesHelper(curr_page.value(), centrality_map, pages);
+    curr_page = q.pop();
+  }
+}
 
-    std::map<std::string, double> delta;
-    for (const auto& page : pages) delta[page] = 0.0;
-    while (!S.empty()) {
-      std::string w = S.top(); S.pop();
-      for (const auto& v : predecessor[w]) {
-        delta[v] += ( sigma[v] / sigma[w] ) * (1.0 + delta[w]);
+void WikiGraph::brandesHelper(const std::string& start, std::map<std::string, double>& centrality_map, const std::vector<std::string>& pages) const {
+  // One run of the Brandes algorithm for a starting node.
+  std::cout << "Thread starting page: " << start << std::endl;
+  std::stack<std::string> S;
+  std::map<std::string, std::vector<std::string>> predecessor;
+  for (const auto& page : pages) predecessor[page] = {};
+  std::map<std::string, double> sigma;
+  for (const auto& page : pages) sigma[page] = 0.0;
+  sigma[start] = 1.0;
+  std::map<std::string, int> dist;
+  for (const auto& page : pages) dist[page] = -1;
+  dist[start] = 0;
+  std::queue<std::string> Q;
+  Q.push(start);
+  while(!Q.empty()) {
+    std::string v = Q.front(); Q.pop();
+    S.push(v);
+    for (const auto& w : article_map.at(v)) {
+      if (dist[w] < 0) {
+        Q.push(w);
+        dist[w] = dist[v] + 1;
       }
-      if (w != start) {
-        centralilty_map[w] += delta[w];
+      if (dist[w] == dist[v] + 1) {
+        sigma[w] = sigma[w] + sigma[v];
+        predecessor[w].push_back(v);
       }
     }
   }
 
-  // centralityMapToFile(centralilty_map, "./output/centrality_map.tsv");
-
-  return centralilty_map; // Default value for now
+  std::map<std::string, double> delta;
+  for (const auto& page : pages) delta[page] = 0.0;
+  while (!S.empty()) {
+    std::string w = S.top(); S.pop();
+    for (const auto& v : predecessor[w]) {
+      delta[v] += ( sigma[v] / sigma[w] ) * (1.0 + delta[w]);
+    }
+    if (w != start) {
+      mtx.lock();
+      centrality_map[w] += delta[w];
```

Comment:
idea: each brandeshelper makes its own map and we store them after each iteration, then add them at the end; this way they don't ever access the same data and we'll just perform $n^2$ add operations in series which is slow but faster than brandes algs in series

Reply:
the thread locking ensures that it isn't accessing the same data. I checked that the data matches, and it at least matches for the most central page (United States). I get where you are coming from, but I think that would be a bit harder to manage, and it hella space-complex.

Reply:
It is possible, but also I am too lazy to try it out.